python/lsst/ts/envsensors/socket_server.py

```diff
@@ -45,119 +47,85 @@ class SocketServer:
+        if server.connected:
+            self.log.info("Starting a new read loop.")
+            self.read_loop_task = asyncio.create_task(self.read_loop())
+
+    async def write(self, data: dict) -> None:
         """Write the data appended with a newline character.
 
         The data are encoded via JSON and then passed on to the StreamWriter
         associated with the socket.
 
         Parameters
         ----------
-        data:
+        data: `dict`
             The data to write.
         """
         self.log.debug(f"Writing data {data}")
         st = json.dumps({**data})
         self.log.debug(st)
-        if self._writer:
-            self._writer.write(st.encode() + b"\r\n")
-            await self._writer.drain()
+        self.writer.write(st.encode() + tcpip.TERMINATOR)
+        await self.writer.drain()
         self.log.debug("Done")
 
-    async def read_loop(self, reader, writer):
-        """Read commands and output replies.
-
-        Parameters
-        ----------
-        reader: stream reader
-            The stream reader to read from.
-        writer: stream writer
-            The stream writer to write to.
-        """
-        self.log.info("The read_loop begins")
-        self._writer = writer
-        self._reader = reader
-
-        while self._started:
-            self.log.debug("Waiting for next incoming message.")
-            try:
-                line = await reader.readuntil(b"\r\n")
-                line = line.decode().strip()
-                self.log.debug(f"Read command line: {line!r}")
-                items = json.loads(line)
-                cmd = items["command"]
-                kwargs = items["parameters"]
-                if cmd == "exit":
-                    await self.exit()
-                else:
-                    await self.command_handler.handle_command(cmd, **kwargs)
-
-            except asyncio.IncompleteReadError:
-                return
-
-    async def exit(self):
+    async def read_loop(self) -> None:
+        """Read commands and output replies."""
+        try:
+            self.log.info(f"The read_loop begins connected? {self.connected}")
+            while self.connected:
+                self.log.debug("Waiting for next incoming message.")
+                try:
+                    line = await self.reader.readuntil(tcpip.TERMINATOR)
+                    line = line.decode().strip()
+                    self.log.debug(f"Read command line: {line!r}")
+                    items = json.loads(line)
+                    cmd = items["command"]
+                    kwargs = items["parameters"]
+                    if cmd == "exit":
+                        await self.exit()
+                    elif cmd == "disconnect":
+                        await self.disconnect()
+                    else:
+                        await self.command_handler.handle_command(cmd, **kwargs)
+
+                except asyncio.IncompleteReadError:
+                    pass
```

Comment:
Please consider moving this `except` up to just after `line = await self.reader...` since none of the rest of this code should raise `asyncio.IncompleteReadError` (I think).

I'm a bit uneasy about the `pass`. It's probably fine but I lean towards `return` so there's no danger of an infinite loop should the connected test be buggy in a way that misses the fact that the connection is lost. Still...I suppose it's conceivable that it'll usefully continue when `return` would not.

Reply:
I have replaced the `pass` with an error log line. Moving the `except` up to where you suggested would involve adding more error handling code so I'll leave it where it is now. A `return` would mean that the loop gets exited which probably is not what should happen. We can discuss this offline if you want.